src/Microsoft.AspNetCore.Blazor/Components/ComponentFactory.cs

```diff
@@ -0,0 +1,104 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+
+namespace Microsoft.AspNetCore.Blazor.Components
+{
+    internal class ComponentFactory
+    {
+        private readonly IServiceProvider _serviceProvider;
+        private readonly static BindingFlags _injectablePropertyBindingFlags
+            = BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic;
+        private readonly IDictionary<Type, Action<IComponent>> _cachedInitializers
+            = new Dictionary<Type, Action<IComponent>>();
+
+        public ComponentFactory(IServiceProvider serviceProvider)
+        {
+            _serviceProvider = serviceProvider
+                ?? throw new ArgumentNullException(nameof(serviceProvider));
+        }
+
+        public IComponent InstantiateComponent(Type componentType)
+        {
+            if (!typeof(IComponent).IsAssignableFrom(componentType))
+            {
+                throw new ArgumentException($"The type {componentType.FullName} does not " +
+                    $"implement {nameof(IComponent)}.", nameof(componentType));
+            }
+
+            var instance = (IComponent)Activator.CreateInstance(componentType);
+            PerformPropertyInjection(instance);
+            return instance;
+        }
+
+        private void PerformPropertyInjection(IComponent instance)
+        {
+            var instanceType = instance.GetType();
+            if (!_cachedInitializers.TryGetValue(instanceType, out var initializer))
+            {
+                initializer = CreateInitializer(instanceType);
+                _cachedInitializers[instanceType] = initializer;
```

Comment:
Is thread safety a concern here?

Reply:
Good point. I was thinking no because all use of a single renderer occurs within the scope of a single thread, and a `ComponentFactory` is owned by a renderer. But maybe in the future our server-side prerenderer may choose to reuse a `ComponentFactory` across renderers (because why wouldn't you) so that would have to be thread-safe. I'll add the necessary locks.

Reply:
Well, I'll make the underlying storage a `ConcurrentDictionary`.